### ISSUE TITLE
Duplicate Trait Objects Warnings

### ISSUE THREAD

**Issue by asquared31415** [rust-lang/rust#82015]
The following code emits two warnings about deprecated trait objects without `dyn`

```rust
#[derive(Clone, Copy)]
struct Double {
    x: Iterator<Item = i32>
}
```

<details><summary>Error</summary>
<pre><code>
warning: trait objects without an explicit `dyn` are deprecated
 --> src/lib.rs:3:8
  |
3 |     x: Iterator<Item = i32>
  |        ^^^^^^^^^^^^^^^^^^^^ help: use `dyn`: `dyn Iterator<Item = i32>`
  |
  = note: `#[warn(bare_trait_objects)]` on by default

warning: trait objects without an explicit `dyn` are deprecated
 --> src/lib.rs:3:8
  |
3 |     x: Iterator<Item = i32>
  |        ^^^^^^^^^^^^^^^^^^^^ help: use `dyn`: `dyn Iterator<Item = i32>`

error[E0204]: the trait `Copy` may not be implemented for this type
 --> src/lib.rs:1:17
  |
1 | #[derive(Clone, Copy)]
  |                 ^^^^
2 | struct Double {
3 |     x: Iterator<Item = i32>
  |     ----------------------- this field does not implement `Copy`
  |
  = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to previous error; 2 warnings emitted
</code></pre>
</details>

`x` can be any trait object without a `dyn`, the important thing is that **both** `Clone` and `Copy` exist, and both can't be implemented (normally for size issues).  I was unable to reproduce using one `derive` and one `impl` on a trait, or two `impl`s.  It seems to be caused specifically by multiple errors in the derive macros.

### Meta
This happens for both debug and release configurations, on all current versions of the compiler.
Tested versions:
Stable 1.50.0 (cb75ad5db 2021-02-10)
Beta 1.51.0-beta.1 (a5a775e3f 2021-02-11)
Nightly 1.52.0-nightly (e9920ef77 2021-02-11)

[Playground Link](https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=9317d361f084fde85a5bbd13ce1d7b17)

**Comment by SkiFire13** [comment]
This code 

```rust
#[derive(Clone, Copy)]
struct Double {
    x: Iterator<Item = i32>
}
```

expands to:

```rust
#![feature(prelude_import)]
#[prelude_import]
use std::prelude::v1::*;
#[macro_use]
extern crate std;
struct Double {
    x: Iterator<Item = i32>,
}
#[automatically_derived]
#[allow(unused_qualifications)]
impl ::core::clone::Clone for Double {
    #[inline]
    fn clone(&self) -> Double {
        {
            let _: ::core::clone::AssertParamIsClone<Iterator<Item = i32>>;
            *self
        }
    }
}
#[automatically_derived]
#[allow(unused_qualifications)]
impl ::core::marker::Copy for Double { }
```

While this code (notice there's no `Copy` derived)

```rust
#[derive(Clone)]
struct Double {
    x: Iterator<Item = i32>
}
```

expands to:

```rust
#![feature(prelude_import)]
#[prelude_import]
use std::prelude::v1::*;
#[macro_use]
extern crate std;
struct Double {
    x: Iterator<Item = i32>,
}
#[automatically_derived]
#[allow(unused_qualifications)]
impl ::core::clone::Clone for Double {
    #[inline]
    fn clone(&self) -> Double {
        match *self {
            Double { x: ref __self_0_0 } =>
            Double{x: ::core::clone::Clone::clone(&(*__self_0_0)),},
        }
    }
}
```

Notice how in the code without `Copy` there's only one usage of `dyn Iterator<Item = i32>` but in the code with `Copy` there are 2, the second one being the one in the `AssertParamIsClone`.

**Comment by asquared31415** [comment]
That makes sense.  I just noticed that in my example warning, the first warning is **just** `Iterator` and no associated item, so it isn't a true duplication of the warnings, though it still is a little odd.

**Comment by asquared31415** [comment]
This was fixed at some point, in 1.63 the output is
```
warning: trait objects without an explicit `dyn` are deprecated
 --> src/lib.rs:3:8
  |
3 |     x: Iterator<Item = i32>
  |        ^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(bare_trait_objects)]` on by default
  = warning: this is accepted in the current edition (Rust 2018) but is a hard error in Rust 2021!
  = note: for more information, see <https://doc.rust-lang.org/nightly/edition-guide/rust-2021/warnings-promoted-to-error.html>
help: use `dyn`
  |
3 -     x: Iterator<Item = i32>
3 +     x: dyn Iterator<Item = i32>
  |

error[[E0204]](https://doc.rust-lang.org/stable/error-index.html#E0204): the trait `Copy` may not be implemented for this type
 --> src/lib.rs:1:17
  |
1 | #[derive(Clone, Copy)]
  |                 ^^^^
2 | struct Double {
3 |     x: Iterator<Item = i32>
  |     ----------------------- this field does not implement `Copy`
  |
  = note: this error originates in the derive macro `Copy` (in Nightly builds, run with -Z macro-backtrace for more info)
  ```
